### PR TITLE
udev: add rules to enable PCIe ASPM for links behind Intel VMD

### PR DESCRIPTION
**Pull request by khfeng**
Intel Tiger Lake platform is equipped with Intel Volume Management
Device, which remaps PCIe root ports and NVMe drives behind another PCI
domain.

Unfortunately, BIOS doesn't enable ASPM for PCIe links behind VMD, makes
battery hours miserable.

However, kernel config "CONFIG_PCIEASPM_DEFAULT=y", which is shipped by
most distros, means "Use the BIOS defaults", so we can't disregard BIOS
default and enable ASPM from the driver.

Use a new udev rules to enable PCIe ASPM for links behind Intel VMD.

### CODE REVIEW

File: rules.d/60-aspm.rules [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION!="add", GOTO="aspm_end"
```
Comment (bluca) [inline rules.d/60-aspm.rules]: shouldn't this be ACTION=="remove" as per https://github.com/systemd/systemd/blob/v247/NEWS#L52 ?

Comment (khfeng) [reply rules.d/60-aspm.rules]: @bluca updated, thanks!

**Comment by bluca** [comment]
Looks OK to me but not an udev expert - @keszybz could you please double check as well? Thanks!

**Comment by poettering** [comment]
i don't get this? why is this a udev/userspace problem? Why isn't this done in the kernel drivers? This looks a lot like overriding an explicit decision the kernel maintainers made by setting CONFIG_PCIEASPM_DEFAULT=y, no?

Isn't this something kernel people should figure out themselves?

I have no idea about any of this, I doubt anyone among the systemd people have, and it looks like a cop-out from something the kernel people cannot agree on themselves? Why would we want to maintain that in udev? 

I really don#t grok this.

**Comment by bluca** [comment]
> i don't get this? why is this a udev/userspace problem? Why isn't this done in the kernel drivers? This looks a lot like overriding an explicit decision the kernel maintainers made by setting CONFIG_PCIEASPM_DEFAULT=y, no?
> 
> Isn't this something kernel people should figure out themselves?
> 
> I have no idea about any of this, I doubt anyone among the systemd people have, and it looks like a cop-out from something the kernel people cannot agree on themselves? Why would we want to maintain that in udev?
> 
> I really don#t grok this.

From the PR description, my impression was that this is a workaround not for the kernel config, which sounds like it's fine in general, but for the combination of broken BIOS behaviour and these specific hw devices.

I really know very little about this area - is this something that could be fixed in the kernel? @khfeng if that's the case, could you please bring it up with the relevant kernel devs/maintainers? Thanks!

**Comment by khfeng** [comment]
> i don't get this? why is this a udev/userspace problem?

No it's not a userspace problem.
If there's anyone to blame, it's BIOS to blame.

However, we still need to take care of this for end users.

> Why isn't this done in the kernel drivers?

Because enabling ASPM with "CONFIG_PCIEASPM_DEFAULT=y" breaks the promise on "Use the BIOS defaults".

> This looks a lot like overriding an explicit decision the kernel maintainers made by setting CONFIG_PCIEASPM_DEFAULT=y, no?

In principle, we can use CONFIG_PCIEASPM_POWERSAVE=y to solve the issue. However, no distro ships with that config.

> 
> Isn't this something kernel people should figure out themselves?
> 
> I have no idea about any of this, I doubt anyone among the systemd people have, and it looks like a cop-out from something the kernel people cannot agree on themselves?

Discussions with kernel folks can be found here:
https://lore.kernel.org/linux-pci/20200821123222.32093-1-kai.heng.feng@canonical.com/
https://lore.kernel.org/linux-pci/20200930082455.25613-1-kai.heng.feng@canonical.com/

> Why would we want to maintain that in udev?

The same way `60-autosuspend.rules` is maintained in udev? Arguably autosuspend should also be handled by kernel driver.

> 
> I really don#t grok this.

Hopefully my reply brings more insight.

**Comment by khfeng** [comment]
> > i don't get this? why is this a udev/userspace problem? Why isn't this done in the kernel drivers? This looks a lot like overriding an explicit decision the kernel maintainers made by setting CONFIG_PCIEASPM_DEFAULT=y, no?
> > Isn't this something kernel people should figure out themselves?
> > I have no idea about any of this, I doubt anyone among the systemd people have, and it looks like a cop-out from something the kernel people cannot agree on themselves? Why would we want to maintain that in udev?
> > I really don#t grok this.
> 
> From the PR description, my impression was that this is a workaround not for the kernel config, which sounds like it's fine in general, but for the combination of broken BIOS behaviour and these specific hw devices.

According to computer manufacturers, the VMD ASPM setting is left blank from Intel's reference BIOS. ASPM is enabled by VMD driver on Windows. Not sure it's enabled from kernelspace or userspace on Windows.

> 
> I really know very little about this area - is this something that could be fixed in the kernel? @khfeng if that's the case, could you please bring it up with the relevant kernel devs/maintainers? Thanks!

I already did, see my previous reply for the links.

In addition to VMD, Realtek ethernet driver r8169 also disables ASPM in the driver. I plan to extend this udev rules for r8169 if this gets merged.

**Comment by poettering** [comment]
> > i don't get this? why is this a udev/userspace problem?
> 
> No it's not a userspace problem.
> If there's anyone to blame, it's BIOS to blame.

But usually the place to work around BIOS issues is the kernel, no?
 
> However, we still need to take care of this for end users.
> 
> > Why isn't this done in the kernel drivers?
> 
> Because enabling ASPM with "CONFIG_PCIEASPM_DEFAULT=y" breaks the promise on "Use the BIOS defaults".

Yeah, but what is that promise worth if we override it in userspace? You might as well override in in kernel space then?

> > This looks a lot like overriding an explicit decision the kernel maintainers made by setting CONFIG_PCIEASPM_DEFAULT=y, no?
> 
> In principle, we can use CONFIG_PCIEASPM_POWERSAVE=y to solve the issue. However, no distro ships with that config.

Why not?

> > Why would we want to maintain that in udev?
> 
> The same way `60-autosuspend.rules` is maintained in udev? Arguably autosuspend should also be handled by kernel driver.

We understand USB autosuspend mutch better and it works quite differently there: it's a single knob sourced from hwdb that we apply. There are a large number of devices served by a single driver there of which a good chunk benefit from the extra info our hw database brings to the table.

This one is much more opaque, i have no clue what those rules you apply do, and it is applied to a single device only.

Generally, we don't do vid/pid matches anymore in udev rules upstream anymore. rules files should be generic, and if they need something device specific, then that's something that should be placed in hwdb.

Is this something that is an intel-only issue? Is this something that wall pop up all across the device scape?

udev shouldn't be the place to ship device specific rules for a single vendor device — except if that's in hwdb, and thus can be expressed generically, with the specific stuff in a static db.

**Comment by khfeng** [comment]
> > No it's not a userspace problem.
> > If there's anyone to blame, it's BIOS to blame.
> 
> But usually the place to work around BIOS issues is the kernel, no?

That's true.

> > Because enabling ASPM with "CONFIG_PCIEASPM_DEFAULT=y" breaks the promise on "Use the BIOS defaults".
> 
> Yeah, but what is that promise worth if we override it in userspace? You might as well override in in kernel space then?

True, but this also applies to many other udev rules.

> > In principle, we can use CONFIG_PCIEASPM_POWERSAVE=y to solve the issue. However, no distro ships with that config.
> 
> Why not?

Most distros only provide one kernel binary for one architecture.
Workstations/servers BIOS may disable all ASPM to get out best PCIe performance. Laptops/tablets enable ASPM to have better battery life.

In reality, all of them use the same kernel shipped by distros, so "use BIOS default" strikes a good balance for all scenarios.

> > The same way `60-autosuspend.rules` is maintained in udev? Arguably autosuspend should also be handled by kernel driver.
> 
> We understand USB autosuspend mutch better and it works quite differently there: it's a single knob sourced from hwdb that we apply. There are a large number of devices served by a single driver there of which a good chunk benefit from the extra info our hw database brings to the table.
> 
> This one is much more opaque, i have no clue what those rules you apply do, and it is applied to a single device only.

This one is actually simpler than USB autosuspend. Enable the knob will enable the specified ASPM link state.

> Generally, we don't do vid/pid matches anymore in udev rules upstream anymore. rules files should be generic, and if they need something device specific, then that's something that should be placed in hwdb.
> 
> Is this something that is an intel-only issue? Is this something that wall pop up all across the device scape?

For now, yes, VMD is only used by Intel. I can't say for the future.

> 
> udev shouldn't be the place to ship device specific rules for a single vendor device — except if that's in hwdb, and thus can be expressed generically, with the specific stuff in a static db.

This one is a bit different. We only want to enable ASPM for links between VMD and NVMe, to keep regular NVMe untouched.

**Comment by poettering** [comment]
If this is intel specific, why not ave it in a separate intel specific package?

For something to be shipped along in udev it must be very common, or very generic. This looks like something very specific to one piece of hw by one vendor.

I don't even know what "vmd" is?

**Comment by poettering** [comment]
> In reality, all of them use the same kernel shipped by distros, so "use BIOS default" strikes a good balance for all scenarios.

sure, but why should *we* override it then? If your patch is not a good default it has no place in systemd/udev. We want to ship good defaults.

**Comment by khfeng** [comment]
> If this is intel specific, why not ave it in a separate intel specific package?

Ok, nice suggestion.

> 
> For something to be shipped along in udev it must be very common, or very generic. This looks like something very specific to one piece of hw by one vendor.
> 
> I don't even know what "vmd" is?

Yes this is Intel specific.

@superm1 any suggestion which package we can put this udev rules in?

**Comment by khfeng** [comment]
> > In reality, all of them use the same kernel shipped by distros, so "use BIOS default" strikes a good balance for all scenarios.
> 
> sure, but why should _we_ override it then? If your patch is not a good default it has no place in systemd/udev. We want to ship good defaults.

No, this is a good default, especially for laptop users.

It's a question on whether we should put it in kernel, systemd or some other userspace package.

**Comment by poettering** [comment]
> No, this is a good default, especially for laptop users.

But if it's a *good* default, then this should be what the kernel does out-of-the-box?

> It's a question on whether we should put it in kernel, systemd or some other userspace package.

Why not kernel? everything would be so much better if the kernel would just come up with good defaults by default.

**Comment by khfeng** [comment]
> > No, this is a good default, especially for laptop users.
> 
> But if it's a _good_ default, then this should be what the kernel does out-of-the-box?
> 
> > It's a question on whether we should put it in kernel, systemd or some other userspace package.
> 
> Why not kernel? everything would be so much better if the kernel would just come up with good defaults by default.

Right, let me think this over.

**Comment by superm1** [comment]
>According to computer manufacturers, the VMD ASPM setting is left blank from Intel's reference BIOS. ASPM is enabled by VMD driver on Windows. Not sure it's enabled from kernelspace or userspace on Windows.

And because of this even if Dell fixes it in BIOS, you'll still hit it on Lenovo.  If Lenovo fixes it you'll hit it on Acme system.  So if it's done wrong in reference BIOS it needs to be handled in kernel or udev rule based upon Linux ecosystem with kernel ASPM settings to respect BIOS settings.

> I don't even know what "vmd" is?

https://chipict.com/intel_vmd_vroc/#:~:text=Intel%C2%AE%20Volume%20Management%20Device%20%28VMD%29%20The%20Intel%C2%AE%20Xeon,chipset%2C%20but%20VMD%20is%20what%20actually%20enables%20VROC. explains it a bit.

> Why not kernel? everything would be so much better if the kernel would just come up with good defaults by default.

The argument against kernel is if there is ever a scenario you don't want ASPM defaults to push down.  VMD can be used in laptops, workstations, racks.  If you want to make a policy decision not to turn on ASPM tied to a chassis that will be more difficult to accomplish in the kernel.

**Comment by poettering** [comment]
> The argument against kernel is if there is ever a scenario you don't want ASPM defaults to push down. VMD can be used in laptops, workstations, racks. If you want to make a policy decision not to turn on ASPM tied to a chassis that will be more difficult to accomplish in the kernel.

People can always override the kernel's default via udev rules, that's entirely fine. But that doesn't mean udev has to ship those in the first place. The kernel should always come with good, sensible defaults that work in most cases — even if there is a sizable of cases where the defaults aren't appropriate. And if people want to depart from those defaults then locally, then that's totally fine, and they can add in udev rules to do so. But unless there's major benefit of being able to leverage hwdb or so I don#t see any reason to have those OK defaults in the kerne itself. 

Or to say this differently:
1. The fewer udev rules we have the better. 
2. If the kernel can come up with sane policy that works for 90% of the cases, that's inherently a good thing
3. If the kernel has to contain a static database with lots of boring entries to implement 1 + 2 then that's a bad thing

Given that we are talking about a single device here so far, point 3 above is a non-issue, hence points 1 + 2 clearly point towards just setting things up properly in the kernel.

If one day there are many more devices that need something like this, and thus point 3 above after all becomes an issue we can still revisit things, but until then I am pretty sure what you are asking should just be done in the kernel.

And again: if people want to locally depart from the kernel's defaults, they then can use udev rules for that, but they should write them and drop them in locally, they shouldn't be shipped with udev.

**Comment by poettering** [comment]
Was this address in kernel somehow?

if not, and this is still relevant it might be OK to add this, but it needs to be done properly: vid/pid matches should be in hwdb, and a udev rule then respects the property from there and applies it. But vid/pid matches in udev rules we ship is not OK anymore. Matches on specific hw really needs to be done in hwdb.

That all said this still appears to be something the kernel should carry if it's really a one-off.

**Comment by superm1** [comment]
@khfeng can you update current status from the kernel? ^

File: rules.d/60-aspm.rules [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+LABEL="aspm_enable"
+
+TEST=="link/clkpm", ATTR{link/clkpm}="1"
+TEST=="link/l0s_aspm", ATTR{link/l0s_aspm}="1"
+TEST=="link/l1_aspm", ATTR{link/l1_aspm}="1"
+TEST=="link/l1_1_aspm", ATTR{link/l1_1_aspm}="1"
+TEST=="link/l1_1_pcipm", ATTR{link/l1_1_pcipm}="1"
+TEST=="link/l1_2_aspm", ATTR{link/l1_2_aspm}="1"
+TEST=="link/l1_2_pcipm", ATTR{link/l1_2_pcipm}="1"
```
Comment (keszybz) [inline rules.d/60-aspm.rules:18]: Is this all the links there are? Is this list the same for all (potential) devices? Some kind of glob assignment would probably be better.

**Comment by khfeng** [comment]
@superm1 I'll update both kernel and this PR soon.

**Comment by poettering** [comment]
so, can we close this?

**Comment by khfeng** [comment]
> so, can we close this?

Please don't :)
Let me work on this...

**Comment by poettering** [comment]
so where are we with this? was this ever addressed kernel side, or in some intel-specific package? Can we close this PR, it seems kinda dead?

**Comment by khfeng** [comment]
Sure. I saw Intel is pursuing a kernel side workaround.

**Comment by poettering** [comment]
Excellent, thanks!

**Pull request closed without merging**